Make fails on older compiler

I have been building Raku on Centos 6 since 2015, with only the occasional [issue](https://github.com/MoarVM/MoarVM/issues/910#issuecomment-406275930).

While updating Raku this morning, Moar failed to make with the following error

```
% /usr/tools/rakudo/nqp/MoarVM (git)-[6c7ee4e...] # make
compiling src/main.o
compiling src/core/callsite.o
compiling src/core/args.o
compiling src/core/exceptions.o
src/core/exceptions.c: In function ‘run_handler’:
src/core/exceptions.c:347:9: warning: empty declaration [enabled by default]
         MVM_FALLTHROUGH
         ^
compiling src/core/interp.o
src/core/interp.c: In function ‘MVM_interp_run’:
src/core/interp.c:3317:25: error: ‘for’ loop initial declarations are only allowed in C99 mode
                         for (MVMuint32 i = 0; i < cu->body.num_frames; i++) {
                         ^
src/core/interp.c:3317:25: note: use option -std=c99 or -std=gnu99 to compile your code
make: *** [src/core/interp.o] Error 1
```

By default, Centos 6 is running cc/gcc 4.4.7. I tried enabling 4.8.2 via Developer Toolset (which has previously worked around issue with the older compiler) however it's still failing.

Below is the output of the Configure.pl for more info

```
Welcome to MoarVM!

Updating submodules .................................... OK
Configuring native build environment ...................
    trying to compile a simple C program ............... YES
    probing support of -Werror=* ....................... YES
    Testing pkgconfig .................................. 0.23
did not find libzstd; will not use heap snapshot format version 3
OK

    probing whether your compiler thinks that it is gcc  YES
    probing how your compiler does static inline ....... static __inline__
    your CPU can read unaligned values for all of int32 int64 num64
    probing the size of pointers ....................... 8
    probing C type support for: _Bool, bool ............ YES: _Bool,bool
    probing computed goto support ...................... YES
    probing pthread_yield support ...................... NO
    probing existance of optional malloc_trim() ........ YES
    probing support of rdtscp intrinsic ................ YES

        make: make
     compile: gcc -Wextra -Wall -Wno-unused-parameter -Wno-unused-function -Wno-missing-braces -Werror=pointer-arith -O3 -DNDEBUG  -D_REENTRANT -D_FILE_OFFSET_BITS=64 -fPIC -DMVM_HEAPSNAPSHOT_FORMAT=2
    includes:   -I3rdparty/libuv/include -I3rdparty/libuv/src -I3rdparty/libatomicops/src -I3rdparty/libtommath -I3rdparty/dyncall/dynload -I3rdparty/dyncall/dyncall -I3rdparty/dyncall/dyncallback
        link: gcc  -O3 -DNDEBUG -Wl,-rpath,"/@libdir@"
        libs:  -lm -lpthread -lrt -ldl

  byte order: little endian

Configuring 3rdparty libs .............................. OK

  3rdparty: 3rdparty/cmp/libcmp.a
            3rdparty/dyncall/dyncall/libdyncall_s.a
            3rdparty/dyncall/dyncallback/libdyncallback_s.a
            3rdparty/dyncall/dynload/libdynload_s.a
            3rdparty/libatomicops/src/libatomic_ops.a
            3rdparty/libtommath/libtommath.a
            3rdparty/libuv/libuv.a
            3rdparty/sha1/libsha1.a
            3rdparty/tinymt/libtinymt.a

Generating src/gen/config.c ............................ OK
Generating src/gen/config.h ............................ OK
Generating Makefile .................................... OK
Generating tools/check.mk .............................. OK
Generating build/mk-moar-pc.pl ......................... OK

Configuration SUCCESS.
```

FYI, I bisected this to commit [520c351](520c351244366c3b60c88b2f37e71db39046641f)

As far as compiler options go, despite the error saying to use `c99` or `gnu99`, it stil fails on `c99`, albeit with a different error.

```
% CFLAGS='-std=c99' perl Configure.pl
Welcome to MoarVM!

Updating submodules .................................... OK
Configuring native build environment ...................
    trying to compile a simple C program ............... YES
    probing support of -Werror=* ....................... YES
    Testing pkgconfig .................................. 0.23
did not find libzstd; will not use heap snapshot format version 3
OK

    probing whether your compiler thinks that it is gcc  YES
    probing how your compiler does static inline ....... static __inline__
    your CPU can read unaligned values for all of int32 int64 num64
    probing the size of pointers ....................... 8
    probing C type support for: _Bool, bool ............ YES: _Bool,bool
    probing computed goto support ...................... YES
    probing pthread_yield support ...................... NO
    probing existance of optional malloc_trim() ........ YES
    probing support of rdtscp intrinsic ................ YES

        make: make
     compile: gcc -Wextra -Wall -Wno-unused-parameter -Wno-unused-function -Wno-missing-braces -Werror=pointer-arith -O3 -DNDEBUG  -D_REENTRANT -D_FILE_OFFSET_BITS=64 -fPIC -DMVM_HEAPSNAPSHOT_FORMAT=2 -std=c99
    includes:   -I3rdparty/libuv/include -I3rdparty/libuv/src -I3rdparty/libatomicops/src -I3rdparty/libtommath -I3rdparty/dyncall/dynload -I3rdparty/dyncall/dyncall -I3rdparty/dyncall/dyncallback
        link: gcc  -O3 -DNDEBUG -Wl,-rpath,"/@libdir@"
        libs:  -lm -lpthread -lrt -ldl

  byte order: little endian

Configuring 3rdparty libs .............................. OK

  3rdparty: 3rdparty/cmp/libcmp.a
            3rdparty/dyncall/dyncall/libdyncall_s.a
            3rdparty/dyncall/dyncallback/libdyncallback_s.a
            3rdparty/dyncall/dynload/libdynload_s.a
            3rdparty/libatomicops/src/libatomic_ops.a
            3rdparty/libtommath/libtommath.a
            3rdparty/libuv/libuv.a
            3rdparty/sha1/libsha1.a
            3rdparty/tinymt/libtinymt.a

Generating src/gen/config.c ............................ OK
Generating src/gen/config.h ............................ OK
Generating Makefile .................................... OK
Generating tools/check.mk .............................. OK
Generating build/mk-moar-pc.pl ......................... OK

Configuration SUCCESS.

Type 'make' to build and 'make help' to see a list of
available make targets.
% make
compiling src/main.o
In file included from 3rdparty/libuv/include/uv.h:66:0,
                 from src/moar.h:19,
                 from src/main.c:4:
3rdparty/libuv/include/uv/unix.h:136:1: error: unknown type name ‘pthread_rwlock_t’
 typedef pthread_rwlock_t uv_rwlock_t;
 ^
In file included from src/moar.h:19:0,
                 from src/main.c:4:
3rdparty/libuv/include/uv.h:330:42: warning: ‘struct addrinfo’ declared inside parameter list [enabled bydefault]
                                   struct addrinfo* res);
                                          ^
make: *** [src/main.o] Error 1
```

It does succeed if I set `CFLAGS='-std=gnu99'`. I still get a lot of warnings, eg.

```
src/core/exceptions.c:347:9: warning: empty declaration [enabled by default]
         MVM_FALLTHROUGH
         ^
compiling src/core/interp.o
src/core/interp.c: In function ‘MVM_interp_run’:
src/core/interp.c:3323:66: warning: ‘coderef’ may be used uninitialized in this function [-Wmaybe-uninitialized]
                         GET_REG(cur_op, 0).o = coderef ? coderef : cu->body.coderefs[0];
                                                                  ^
src/core/interp.c:3316:36: warning: variable ‘coderef’ might be clobbered by ‘longjmp’ or ‘vfork’ [-Wclobbered]
                         MVMObject *coderef;

...

src/io/signals.c:119:3: warning: ‘MVM_sig_names’ defined but not used [-Wunused-variable]
 } MVM_sig_names;
   ^

...

src/strings/utf8_c8.c:134:18: warning: ‘*((void *)&encoded+3)’ may be used uninitialized in this function[-Wmaybe-uninitialized]
         MVMuint8 low = invalid & 0x0F;
```

... but these warnings we're happening prior to commit [520c351](520c351244366c3b60c88b2f37e71db39046641f) and in the end it appears to build ok with `gnu99`.

So... yes I suppose I could modify the build args to use `gnu99`, but I'm not sure what the repercussions of that could be to other users, and thought it might be of more interest to understand what exactly in that commit is causing the build failure.

Personally, I'm willing to work around it manually, knowing that CentOS6/RHEL6 is quite old and is reaching EOL this year, so it likely doesn't affect many Raku users.

> Does that work for you?

Running `CFLAGS='--std=c99 -D_DEFAULT_SOURCE' perl Configure.pl` and then `make` fails with the same errors as shown above when I tried to build with just  `CFLAGS='--std=c99`.

ie.
```
3rdparty/libuv/include/uv/unix.h:136:1: error: unknown type name ‘pthread_rwlock_t’
 typedef pthread_rwlock_t uv_rwlock_t;
...
```

So going back and looking at the original error...

```
src/core/interp.c:3317:25: error: ‘for’ loop initial declarations are only allowed in C99 mode
                         for (MVMuint32 i = 0; i < cu->body.num_frames; i++) {
                         ^
```

[This](https://github.com/MoarVM/MoarVM/blob/520c351244366c3b60c88b2f37e71db39046641f/src/core/interp.c#L3317) seems to be the main offender.

Build appears to work again on Centos 6 after making the following change

```
3317c3317,3318
<                         for (MVMuint32 i = 0; i < cu->body.num_frames; i++) {
---
>                         MVMuint32 i;
>                         for (i = 0; i < cu->body.num_frames; i++) {
```

--- a/src/core/interp.c
+++ b/src/core/interp.c
@@ -3313,7 +3313,8 @@ void MVM_interp_run(MVMThreadContext *tc, void (*initial_invoke)(MVMThreadContex
                     MVMCompUnit *cu = (MVMCompUnit *)maybe_cu;
                     if (cu->body.mainline_frame) {
                         MVMObject *coderef;
-                        for (MVMuint32 i = 0; i < cu->body.num_frames; i++) {
+                        MVMuint32 i;
+                        for (i = 0; i < cu->body.num_frames; i++) {
                             if (((MVMCode*)cu->body.coderefs[i])->body.sf == cu->body.mainline_frame) {
                                 coderef = cu->body.coderefs[i];
                                 break;

